[CI] FieldExtractorIT class failing

**Build Scans:**
- [elasticsearch-intake #43134 / 9.3.8_bwc-snapshots](https://gradle-enterprise.elastic.co/s/a22yvostyws64)
- [elasticsearch-pull-request #162073 / 8.19.19_part_3_bwc-snapshots](https://gradle-enterprise.elastic.co/s/ffwl2gozbadpi)
- [elasticsearch-pull-request #161778 / 9.3.8_part_3_bwc-snapshots](https://gradle-enterprise.elastic.co/s/7dxqjb3grjn32)

**Reproduction Line:**
```
undefined
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.esql.qa.mixed.FieldExtractorIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('org.elasticsearch.xpack.esql.qa.mixed.FieldExtractorIT'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
undefined
```

**Issue Reasons:**
- [main] 2 consecutive failures in class org.elasticsearch.xpack.esql.qa.mixed.FieldExtractorIT
- [main] 1 consecutive failures in step 9.3.8_bwc-snapshots
- [main] 1 consecutive failures in step 8.19.19_part_3_bwc-snapshots
- [main] 1 consecutive failures in pipeline elasticsearch-intake
- [main] 1 consecutive failures in pipeline elasticsearch-pull-request
- [main] 3 failures in class org.elasticsearch.xpack.esql.qa.mixed.FieldExtractorIT (0.3% fail rate in 1000 executions)
- [main] 2 failures in pipeline elasticsearch-pull-request (1.0% fail rate in 208 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

Found these clue in the test log

```
  x-pack/plugin/esql/qa/server/mixed-cluster/build/test-results/v9.3.8#javaRestTest/TEST-org.elasticsearch.xpack.esql.qa.mixed.FieldExtractorIT.xml

  1. Fatal crash — test-cluster-0 (9.3.8, the master), in <system-out> (line 1095)

  [2026-07-08T15:34:38,411][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [test-cluster-0] fatal error in thread
  [elasticsearch[test-cluster-0][masterService#updateTask][T#1]], exiting
  java.lang.AssertionError: expected [index.time_series.temporality_field] to be private but it was not
      at org.elasticsearch.server@9.3.8-SNAPSHOT/...MetadataCreateIndexService.validatePrivateSettingsNotExplicitlySet(MetadataCreateIndexService.java:1638)
      at org.elasticsearch.server@9.3.8-SNAPSHOT/...MetadataCreateIndexService.getIndexSettingsValidationErrors(MetadataCreateIndexService.java:1624)
      at org.elasticsearch.server@9.3.8-SNAPSHOT/...MetadataIndexTemplateService.validate(MetadataIndexTemplateService.java:2185)
      ...
      at org.elasticsearch.server@9.3.8-SNAPSHOT/...MetadataIndexTemplateService.addIndexTemplateV2(MetadataIndexTemplateService.java:740)
      at org.elasticsearch.server@9.3.8-SNAPSHOT/...MasterService.executeAndPublishBatch(MasterService.java:261)

  2. Fatal crash — test-cluster-2 (9.3.8, took over as master), in <system-out> (line 1311)

  [2026-07-08T15:34:39,954][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [test-cluster-2] fatal error in thread
  [elasticsearch[test-cluster-2][masterService#updateTask][T#1]], exiting
  java.lang.AssertionError: expected [index.time_series.temporality_field] to be private but it was not
      at org.elasticsearch.server@9.3.8-SNAPSHOT/...MetadataCreateIndexService.validatePrivateSettingsNotExplicitlySet(MetadataCreateIndexService.java:1638)
      at org.elasticsearch.server@9.3.8-SNAPSHOT/...MetadataCreateIndexService.getIndexSettingsValidationErrors(MetadataCreateIndexService.java:1624)

```
AI assisted analysis

**Root cause**

The cluster is a 4-node mix: nodes -1/-3 are 9.6.0-SNAPSHOT (new) and nodes -0/-2 are 9.3.8-SNAPSHOT (old/BWC). During formation an old 9.3.8 node was elected master, and while installing the built-in stack/otel templates it hit a fatal assertion and exited:

  [test-cluster-0] fatal error ... exiting
  java.lang.AssertionError: expected [index.time_series.temporality_field] to be private but it was not
    at MetadataCreateIndexService.validatePrivateSettingsNotExplicitlySet(MetadataCreateIndexService.java:1638)
    at MetadataIndexTemplateService.validate(...)
    at MetadataIndexTemplateService.addIndexTemplateV2(...)

Both 9.3.8 nodes crashed the same way (node -0, then -2 took over as master and crashed too). Losing 2 of 4 master-eligible nodes broke quorum, so the cluster could never become ready → waitUntilReady → connection refused.

The assertion that fires (this is 9.3.8 code):
```
   server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java lines 1883-1887

          for (final String key : settings.keySet()) {
              final Setting<?> setting = indexScopedSettings.get(key);
              if (setting == null) {
                  assert indexScopedSettings.isPrivateSetting(key) : "expected [" + key + "] to be private but it was not";
              } else if (setting.isPrivateIndex() && isSystemProvided(key, settings, systemProvided) == false) {
```
On the 9.3.8 node, index.time_series.temporality_field is an unknown setting (indexScopedSettings.get(...) == null) and is not a private setting, so the assertion trips.

How the setting got into a built-in template

The setting is new (defined in IndexSettings.TIME_SERIES_TEMPORALITY_FIELD, added on main/9.6.0 by #144912, never backported to 9.3). It became part of the metrics-otel@mappings component template — pulled in by the composable template metrics-otel@template (composed_of: ... - metrics-otel@mappings) — via PR #152091 (commit 86b789b3433, 2026-06-25) "Remove feature flag for TSDB metric temporality support":

```
     settings:
       index.mapping.ignore_malformed: true
       index.mapping.total_fields.limit: 10000
  +    index.time_series.temporality_field: temporality
```

Before that PR the setting was behind TIME_SERIES_TEMPORALITY_FEATURE_FLAG and was not in any shipped template, so old nodes never saw it. After the flag removal, the otel template unconditionally carries a setting that 9.3.x nodes don't recognize.

**Why it's intermittent**

It only fails when a 9.3.8 node wins the master election during formation and validates the otel template. If a 9.6.0 node stays master, the template validates fine and the cluster starts. Note also the crash is an assertion (-ea is on in tests); in production (assertions off) the old master would silently skip validation rather than die — but it's still a genuine mixed-version gap.

**Suggested fix directions (not applied)**

The real issue is introducing a new index setting into a shipped system template in the same release without it being understood by supported older nodes in a rolling upgrade.

Options the owning team (TSDB / otel-data) would typically consider:
1. Gate the index.time_series.temporality_field line in metrics-otel@mappings so it's only installed once the whole cluster supports it (min transport/node version / feature check in the otel template registry), or
2. Backport the setting registration to the active 9.3.x/9.4.x/9.5.x BWC branches so older nodes recognize it, or
3. Harden validatePrivateSettingsNotExplicitlySet so an unknown (non-private) setting doesn't assertion-crash the master.

I also did a detailed log analysis now and came to the conclusion that the root cause isn't the temporality feature.
Instead, it jsut surfaces a rare concurrency problem in the cluster startup and master node election.

The key aspect for the failure is that we try to install the new 9.6 OTel template using the new index setting on a 9.3 master.
This normally shouldn't happen, because `YamlTemplateRegistry` (on which the OTel template registry is based) overrides [`requiresMasterNode()`](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/YamlTemplateRegistry.java#L147) with `true`. This is supposed to ensure that the template is only installed if the node itself is the master (and hence is on 9.6 in our case).

However, the logs proof that `requiresMasterNode()` isn't fully safe, as there is a sequence of events which can lead to the template being installed on an older, remote master.
Let's look at the cluster roles over time:

Initial cluster state
| Node | Version | Node ID |
  |---|---|---|
  | test-cluster-0 | 9.3.8 | Eb5LrWmwQ8O6fFkTGWI1kQ |
  | test-cluster-1 | 9.6.0 | _sX7uDU3Qoyzg3DKUNsLJQ |
  | test-cluster-2 | 9.3.8 | Wl4lvmmXRj6IxnuV6bSOOA |
  | test-cluster-3 | 9.6.0 | B0fmh8B8R7yn8BmyKt0o6g |

  All 4 nodes start simultaneously at `23:19:38`.

events over time

  | Time | Term | Node | Event |
  |---|---|---|---|
  | 23:20:08.067 | 1 | test-cluster-3 (9.6.0) | **Elected master** with 3 nodes (0, 2, 3). test-cluster-1 hasn't joined yet. |
  | 23:20:08.148 | 1 | test-cluster-0 (9.3.8) | Applies cluster state: master = test-cluster-3 |
  | 23:20:08.149 | 1 | test-cluster-2 (9.3.8) | Applies cluster state: master = test-cluster-3 |
  | 23:20:08.194 | 1 | test-cluster-3 (9.6.0) | Applies own cluster state: master = self |
  | 23:20:08.661–09.643 | 1 | test-cluster-3 (9.6.0) | **Installs all component + index templates** as master (including `metrics-otel@mappings` with `temporality_field`, all APM templates, etc.) |
  | 23:20:09.937 | 1 | test-cluster-1 (9.6.0) | Finally applies cluster state v6: master = test-cluster-3. (Joined late — ~2s after the others.) |
  | 23:20:10.007 | 1 | test-cluster-3 (9.6.0) | **Loses master**: "becoming candidate: joinLeaderInTerm" — another node started a new election by sending a join with a higher term. |
  | 23:20:10.019–.027 | — | test-cluster-3 (9.6.0) | 4× `NotMasterException` for pending master tasks (license generation). |
  | 23:20:10.098 | 1 | test-cluster-2 (9.3.8) | Master changed to `[]` (no master): "becoming candidate: joinLeaderInTerm" |
  | 23:20:10.103 | 1 | test-cluster-0 (9.3.8) | Master changed to `[]`: "becoming candidate: joinLeaderInTerm" |
  | 23:20:10.109 | 1 | test-cluster-1 (9.6.0) | Master changed to `[]`: "becoming candidate: joinLeaderInTerm" |
  | 23:20:10.118 | **4** | test-cluster-3 (9.6.0) | **Re-elected master** with 2 nodes (3, 2). Publication starts. |
  | 23:20:10.197 | **6** | test-cluster-1 (9.6.0) | **Elected master** with 3 nodes (0, 1, 2). Higher term supersedes term 4. |
  | 23:20:10.223 | **7** | test-cluster-2 (9.3.8) | **Elected master** with 2 nodes (0, 2) + test-cluster-1 rejoining. Term 7 supersedes all. |
  | 23:20:10.223 | 4 | test-cluster-3 (9.6.0) | **Publication of term-4 cluster state fails** — "failed to commit cluster state version [7] in term [4]". test-cluster-3's second mastership is dead. |
  | 23:20:10.552 | 7 | test-cluster-2 (9.3.8) | Applies own cluster state: master = self (test-cluster-2, 9.3.8). |
  | 23:20:10.776–.809 | 7 | test-cluster-2 (9.3.8) | All other nodes join (0, 1, 3). Cluster is now stable under test-cluster-2. |
  | 23:20:11.054–.555 | 7 | test-cluster-2 (9.3.8) | **Processes forwarded template PUTs**: .kibana-reporting, APM index templates, etc. These are requests **forwarded from 9.6.0 nodes** whose `TransportMasterNodeAction` retry loop detected test-cluster-2 as the new master. |
  | 23:20:11.597 | 7 | test-cluster-2 (9.3.8) | **CRASH**: `AssertionError: expected [index.time_series.temporality_field] to be private but it was not` — while processing a composable index template that references `metrics-otel@mappings` (already in cluster state from test-cluster-3's earlier installation).
  |

So initially, the master node is a 9.6 node. This already is a non-recommended state, as we also have 9.3 nodes in the cluster and we recommend updating the master last ([docs](https://www.elastic.co/docs/deploy-manage/upgrade/deployment-or-cluster/elasticsearch#upgrade-order)).
This alone doesn't cause the problem: The 9.6 would validate the template and wouldn't find any problems, as it knows the new index setting.
The problem is the following sequence of events:
 * The 9.6 master loads the templates. To do so, it submits a cluster state update task to the `MasterService`
 * Before the cluster state update is processed, the 9.6 node looses its master status to a 9.3 node
 * The `MasterService` on the 9.6 node now attempts to process the cluster state update. It fails because the node is no longer the master
 * The cluster state update enters a retry code path where the update is attempted on the new master - the 9.3 node. On the 9.3 node, the assertion fires due to the unknown cluster setting.

 This proves that `requiresMasterNode()` doesn't always protect from accidentally performing the index template upload only to the local node as master.

--- a/docs/changelog/154931.yaml
+++ b/docs/changelog/154931.yaml
@@ -2,5 +2,5 @@ area: TSDB
 issues:
  - 153331
 pr: 154931
-summary: Gate temporality templates on node support
+summary: Gate templates depending on temporality support on cluster feature
 type: bug

--- a/docs/changelog/154931.yaml
+++ b/docs/changelog/154931.yaml
@@ -1,0 +1,6 @@
+area: TSDB
+issues:
+ - 153331
+pr: 154931
+summary: Gate temporality templates on node support
+type: bug

--- a/server/src/main/java/org/elasticsearch/index/mapper/MapperFeatures.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/MapperFeatures.java
@@ -138,6 +138,11 @@ public class MapperFeatures implements FeatureSpecification {
     public static final NodeFeature TSDB_METRIC_TEMPORALITY_SUPPORT = new NodeFeature("mapper.tsdb.metric_temporality_support");
 
     @Override
+    public Set<NodeFeature> getFeatures() {
+        return Set.of(TSDB_METRIC_TEMPORALITY_SUPPORT);
+    }
+
+    @Override
     public Set<NodeFeature> getTestFeatures() {
         return Set.of(
             RangeFieldMapper.DATE_RANGE_INDEXING_FIX,
@@ -226,8 +231,7 @@ public class MapperFeatures implements FeatureSpecification {
             DOC_VALUES_MULTI_VALUE_INDEX_SETTING,
             DOC_VALUES_MULTI_VALUE_FALSE_ALIAS,
             DOC_VALUES_EXTENDED_FORM_ONLY_IN_COLUMNAR,
-            DOC_VALUES_NULLABILITY,
-            TSDB_METRIC_TEMPORALITY_SUPPORT
+            DOC_VALUES_NULLABILITY
         );
     }
 }

--- a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java
+++ b/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java
@@ -43,6 +43,8 @@ import org.elasticsearch.core.TimeValue;
 import org.elasticsearch.features.FeatureService;
 import org.elasticsearch.features.NodeFeature;
 import org.elasticsearch.gateway.GatewayService;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.mapper.MapperFeatures;
 import org.elasticsearch.ingest.IngestMetadata;
 import org.elasticsearch.ingest.PipelineConfiguration;
 import org.elasticsearch.threadpool.ThreadPool;
@@ -121,7 +123,12 @@ public abstract class IndexTemplateRegistry implements ClusterStateListener {
      * This map matches node features with predicates to detect when a template requires that feature.
      * This allows the registry to install the template only after the cluster fully supports a feature.
      */
-    protected static final Map<NodeFeature, Predicate<Template>> NODE_FEATURE_FILTERS = Map.of();
+    protected static final Map<NodeFeature, Predicate<Template>> NODE_FEATURE_FILTERS = Map.of(
+        MapperFeatures.TSDB_METRIC_TEMPORALITY_SUPPORT,
+        template -> template != null
+            && template.settings() != null
+            && template.settings().hasValue(IndexSettings.TIME_SERIES_TEMPORALITY_FIELD.getKey())
+    );
     private final Map<NodeFeature, Predicate<Template>> nodeFeatureFilters;
     // This flag short-circuits the node feature check, if all node features are supported.
     private volatile boolean allFeaturesSupported = false;

--- a/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistryTests.java
+++ b/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistryTests.java
@@ -41,6 +41,8 @@ import org.elasticsearch.core.Strings;
 import org.elasticsearch.features.FeatureService;
 import org.elasticsearch.features.NodeFeature;
 import org.elasticsearch.index.Index;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.mapper.MapperFeatures;
 import org.elasticsearch.ingest.IngestMetadata;
 import org.elasticsearch.ingest.PipelineConfiguration;
 import org.elasticsearch.tasks.Task;
@@ -84,6 +86,7 @@ import static org.hamcrest.Matchers.empty;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.hasSize;
 import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.Matchers.oneOf;
 import static org.hamcrest.Matchers.sameInstance;
 import static org.hamcrest.Matchers.startsWith;
@@ -1110,6 +1113,23 @@ public class IndexTemplateRegistryTests extends ESTestCase {
             containsInAnyOrder("test-index-one@template", "test-index-two@template")
         );
         assertThat(registry.allFeaturesSupported(), equalTo(true));
+    }
+
+    public void testTemporalitySettingRequiresClusterFeature() {
+        Predicate<Template> filter = IndexTemplateRegistry.NODE_FEATURE_FILTERS.get(MapperFeatures.TSDB_METRIC_TEMPORALITY_SUPPORT);
+        assertThat(filter, notNullValue());
+        assertThat(
+            filter.test(
+                new Template(
+                    Settings.builder().put(IndexSettings.TIME_SERIES_TEMPORALITY_FIELD.getKey(), "temporality").build(),
+                    null,
+                    null
+                )
+            ),
+            equalTo(true)
+        );
+        assertThat(filter.test(new Template(Settings.EMPTY, null, null)), equalTo(false));
+        assertThat(filter.test(null), equalTo(false));
     }
 
     public void testFeatureAbsentFiltersMatchingComponentTemplate() {

